File: Day231/reverseLinkedListII.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseBetween(head *ListNode, left int, right int) *ListNode {
    temp := &ListNode{Next: head}
    first := temp
    for i:=1; i<left; i++ {
        first = first.Next
    }
    prev, curr, next := first, first.Next, first.Next.Next
    for i:=left; i<right; i++ {
        prev, curr, next = curr, next, next.Next
        curr.Next = prev
    }
    first.Next.Next = next
    first.Next = curr
    return temp.Next
}

File: Day231/reverseLinkedListII.java
package Day231;
//   Definition for singly-linked list.
   class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
  }
 
class Solution {
    public ListNode reverse(ListNode head){
        ListNode current = head;
        ListNode nextNode = head;
        ListNode prev = null;
        while(current!=null){
            nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }
        return prev;
    }
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode current = head;
        ListNode start = null;
        ListNode end = null;
        int i = 1;
        while(current!=null){
            if(i>right){
                break;
            }
            if(i<left){
                start = current;
            }
            if(i==right){
                end = current;
            }
            current = current.next;
            i++;
        }

        end.next = null;
        if(left == 1){
            end = head;
            head = reverse(head);
        }else{
            end = start.next;
            start.next = reverse(start.next);
        }
        end.next = current;
        return head;

    }
}